e2e/nomostest/reset.go

```diff
@@ -0,0 +1,563 @@
+	repoSyncList, err := listRepoSyncs(nt)
+	if err != nil {
+		return err
+	}
+	if err := ResetRepoSyncs(nt, repoSyncList.Items); err != nil {
+		return err
+	}
+
+	// Delete all Namespaces with the test label (except shared).
+	nsList, err := listNamespaces(nt)
+	if err != nil {
+		return err
+	}
+	// Error if any protected namespace was modified by a test (test label added)
+	// and not reverted by the test.
+	protectedNamespaces, nsListItems := filterNamespaces(nsList.Items,
+		protectedNamespaceList()...)
+	if len(protectedNamespaces) > 0 {
+		return errors.Errorf("protected namespace(s) modified by test: %+v",
+			protectedNamespaces)
+	}
+	// Skip deleting namespaces that contain test infra or Config Sync itself.
+	_, nsListItems = filterNamespaces(nsListItems,
+		configsync.ControllerNamespace,
+		configmanagement.RGControllerNamespace,
+		metrics.MonitoringNamespace,
+		testGitNamespace)
+	if err := ResetNamespaces(nt, nsListItems); err != nil {
+		return err
+	}
+
+	// NOTE: These git repos are not actually being deleted here, just forgotten.
+	// Repos are deleted by `Clean` for environment setup and teardown.
+
+	nt.T.Log("[RESET] Resetting NonRootRepos")
+	nt.NonRootRepos = make(map[types.NamespacedName]*Repository)
+
+	nt.T.Log("[RESET] Resetting RootRepos")
+	// Retain the root-sync repo (speeds up reset a bit when there are no changes)
+	rootRepo := nt.RootRepos[configsync.RootSyncName]
+	nt.RootRepos = make(map[string]*Repository)
+	nt.RootRepos[configsync.RootSyncName] = rootRepo
+
+	return nil
+}
+
+func protectedNamespaceList() []string {
+	list := make([]string, 0, len(differ.SpecialNamespaces))
+	for ns := range differ.SpecialNamespaces {
+		list = append(list, ns)
+	}
+	return list
+}
+
+type resetRecord struct {
+	Objects          map[types.NamespacedName]struct{}
+	ManagedObjects   map[types.NamespacedName]struct{}
+	ObjectNamespaces map[string]struct{}
+}
+
+func newResetRecord() *resetRecord {
+	return &resetRecord{
+		Objects:          make(map[types.NamespacedName]struct{}),
+		ManagedObjects:   make(map[types.NamespacedName]struct{}),
+		ObjectNamespaces: make(map[string]struct{}),
+	}
+}
+
+// ResetRootSyncs cleans up one or more RepoSyncs and all their managed objects.
+// Use this for cleaning up RepoSyncs in tests that use delegated control.
+func ResetRootSyncs(nt *NT, rsList []v1beta1.RootSync) error {
+	if len(rsList) == 0 {
+		return nil
+	}
+
+	nt.T.Logf("[RESET] Deleting RootSyncs (%d)", len(rsList))
+
+	record := newResetRecord()
+
+	for _, item := range rsList {
+		rs := &item
+		rsNN := client.ObjectKeyFromObject(rs)
+		record.Objects[rsNN] = struct{}{}
+		record.ObjectNamespaces[rsNN.Namespace] = struct{}{}
+
+		if manager, found := rs.GetAnnotations()[string(metadata.ResourceManagerKey)]; found {
+			record.ManagedObjects[rsNN] = struct{}{}
+			nt.T.Logf("[RESET] RootSync %s managed by %q", rsNN, manager)
+			if !IsDeletionPropagationEnabled(rs) {
```

Comment:
I'm not opposed to failing hard here, but would it be possible to use server side apply to set the annotation even though it's managed by another `RootSync`? I suppose it might get blocked by `admission-webhook`

Reply:
The admission webhook would deny it and even if it’s disabled, the remediator would revert it, causing a race condition. 

So I just made it hard error so we can fix the test if we see it.